indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java

```diff
@@ -0,0 +1,270 @@
+                final String supervisorTaskId = supervisorTaskDir.getName();
+                supervisorTaskCheckTimes.computeIfAbsent(
+                    supervisorTaskId,
+                    k -> DateTimes.nowUtc().plus(intermediaryPartitionTimeout)
+                );
+              }
+            }
+          }
+        },
+        intermediaryPartitionDiscoveryPeriodSec,
+        intermediaryPartitionDiscoveryPeriodSec,
+        TimeUnit.SECONDS
+    );
+    // Check supervisorTask status if its partitions have not been accessed in timeout.
+    // Delete partitions if the supervisorTask is already finished.
+    // Note that the overlord sends a cleanup request when a supervisorTask is finished. The below check is to trigger
+    // the self-cleanup for when middleManager misses the cleanup request from the overlord.
+    supervisorTaskChecker.scheduleAtFixedRate(
+        () -> {
+          final DateTime now = DateTimes.nowUtc();
+          final Set<String> expiredSupervisorTasks = new HashSet<>();
+          for (Entry<String, DateTime> entry : supervisorTaskCheckTimes.entrySet()) {
+            final String supervisorTaskId = entry.getKey();
+            final DateTime checkTime = entry.getValue();
+            if (checkTime.isAfter(now)) {
+              expiredSupervisorTasks.add(supervisorTaskId);
+            }
+          }
+
+          final Map<String, TaskStatus> taskStatuses = indexingServiceClient.getTaskStatuses(expiredSupervisorTasks);
+          RuntimeException exception = null;
+          for (Entry<String, TaskStatus> entry : taskStatuses.entrySet()) {
+            final String supervisorTaskId = entry.getKey();
+            final TaskStatus status = entry.getValue();
+            if (status.getStatusCode().isComplete()) {
+              try {
+                deletePartitions(supervisorTaskId);
+              }
+              catch (IOException e) {
+                if (exception == null) {
+                  exception = new RuntimeException(e);
+                } else {
+                  exception.addSuppressed(e);
+                }
+              }
+            }
+          }
+
+          if (exception != null) {
+            throw exception;
+          }
+        },
+        intermediaryPartitionCleanupPeriodSec,
+        intermediaryPartitionCleanupPeriodSec,
+        TimeUnit.SECONDS
+    );
+  }
+
+  @LifecycleStop
+  public void stop() throws InterruptedException
+  {
+    if (supervisorTaskChecker != null) {
+      supervisorTaskChecker.shutdownNow();
+      supervisorTaskChecker.awaitTermination(10, TimeUnit.SECONDS);
+      supervisorTaskChecker = null;
+    }
+    supervisorTaskCheckTimes.clear();
+  }
+
+  /**
+   * Write a segment into one of configured locations. The location to write is chosen in a round-robin manner per
+   * supervisorTaskId.
+   *
+   * This method is only useful for the new Indexer model, and must not be called when tasks are running in the existing
```

Comment:
does this mean parallel index task with shuffle support would not work with existing MMs?

why can't task processes have intermediary segment location propagated to them so that they could write data there and whole thing works for existing MMs as well ?

Reply:
can you add what that other method is/would-be ? ... my understanding is that this method could be made static and peon process would write data by calling this method as well.

Reply:
Add a new static method which could be used by MMs.